### ISSUE TITLE
Websocket queued EVENT's limit

### ISSUE THREAD

**Issue by alex3d** [crossbario/crossbar#874]
How can I protect crossbar router from malicious subscriber not reading incoming EVENTs?
With default settings crossbar.io router process eats up all available memory to queue EVENTs and I can't find queue limit in docs.

BTW: bonefish - same issue, 
turnpike - different issue - publisher can't publish new EVENTs when one of subscribers not reading it (and network buffers is filled up)

**Comment by oberstet** [comment]
You mean a subscriber that subscribes to topic T (and is authorized to do so), but then fails to read incoming events (fast enough) from it's connection?

**Comment by alex3d** [comment]
Yes

**Comment by oberstet** [comment]
Yep, we should have a knob to control that situation. Means, memory runs away currently (as you figured).

Background: bytes to be sent on a socket get buffered inside Twisted userland [here](https://github.com/twisted/twisted/blob/trunk/src/twisted/internet/abstract.py#L339).

We might check `transport._isSendBufferFull()` (which checks the userland Twisted buffers) and throw away events for the subscriber or kick the subscriber session completely.

Crossbar.io (deliberately, and by design) is not an event queuing system - it has (subscriber independent) event history instead. So above fits our approach.

> Note for Twisted guys: no, producer-consumer pattern is not a solution here, since we don't want to throttle the publisher side just because a single subscriber cannot keep up.

**Comment by oberstet** [comment]
I make it a "bug". Implementation is trivial essentially. We already have ability in node configuration to configure URI specific things (like eg activate event history). We also have ability to configure transports .. probably add an option "max buffered bytes" there. Then do the check above. This should be done for all stream transports we support (TCP, TLS, Unix domain, pipes, ..).

**Comment by oberstet** [comment]
@alex3d does above approach (throwing away events for that subscriber / kicking the subscriber session) work for you?

I can think of only 2 alternatives in general:

* slowing down the publisher (which would be tricky, because it means end-to-end flow control with backpressure arising from multiple destinations - all the subscribers)
* full blown event queuing

The latter is somewhat related to another feature that we once thought about: we do have acknowledged publications (the default is unacknowledged), but we do not have acknowledged event delivery - the latter would mean keeping book of delivery of events _for each subscriber_, and acting accordingly. The price for that is relatively high.

**Comment by meejah** [comment]
I don't think back-pressure on the publisher from *ANY* subscriber is a good idea -- this would be way too open to "abuse" (even unintentional! like if you're on a really slow connection).

A full-blown event-queuing system also seems like something that's out-of-scope for crossbar itself. Perhaps a better way would be to offer integration to existing durable event-queuing systems? I haven't thought this through -- maybe it wouldn't work well at all -- but there are many systems out there that do this, and it seems well worth leveraging one instead of re-inventing this wheel.

I think the best solution here is to provide the two knobs @oberstet implies: how much backlog is acceptable (either as bytes, or *maybe* number of messages?) and what to do when it's hit. For the latter, "stop sending events" or "nuke subscriber" would be the two obvious options. (Also: I think they *should* be options, since I can easily imagine use-cases for each).

Whatever we choose here, we probably want the same options for caller/callee flows as well, right? That is, if the callee can't keep up with a particular caller. This is probably way more rare than in the publisher/subscriber flow (because the caller is Very Likely waiting for an answer), but could still be an issue (e.g. if you have "fire and forget"-type calls where the caller program moves on before getting the answer).

**Comment by meejah** [comment]
Possibly relevant: the Twisted check only considers bytes not sent to the kernel.

...and the default is 65KiB (!!)

**Comment by oberstet** [comment]
@meejah 

Yes, the Twisted buffering is only the userland buffering (kernel TCP buffers is on top), and yes, the default is 64K (https://github.com/twisted/twisted/blob/trunk/src/twisted/internet/abstract.py#L450).

---

> A full-blown event-queuing system also seems like something that's out-of-scope for crossbar itself. Perhaps a better way would be to offer integration to existing durable event-queuing systems?

We do have event history already - what's missing is persistent event history. This will be done using the already integrated LMDB.

We could extend that use of LMDB to implemented per-subscriber queuing - this isn't hard to do, it just has a big price in performance.

There are no plans to integrate with some external queuing system though .. given that we already have LMDB (which totally rocks for various reasons).

---

Rgd similar control for RPC: we already have something here: Call Queueing / Concurrency Control  https://github.com/crossbario/crossbarexamples/tree/master/scaling-microservices

---

All in all, as a first step, I think we should add a low-level control knob at the transport level `max_buffered_bytes` for the maximum bytes buffered in userland before the session / transport connection gets kicked.
